### PR TITLE
FM demodulation and IQ saving

### PR DESCRIPTION
**Pull request by redradist**
## Description
This PR adds possibility to demodulate audio for which demodulation is applied as well as IQ sample for further analyzes

### CODE REVIEW

File: _signal_processing/signal_utils.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+def fm_demod(samples, sample_freq, freq_width):
+    demod_channel = samples
+
+    demod_channel = firwin_lowpass_filter(demod_channel, int(freq_width/2), sample_freq)
+
+    baseband = demod_channel
+    demod_channel = np.angle(baseband[1::1] * np.conjugate(baseband[0:-1:1]))
+    demod_channel = (demod_channel * sample_freq) / (2 * np.pi)
+
```
Comment (godsic) [inline _signal_processing/signal_utils.py:70]: You might need to add deemphasis filter here.

File: _signal_processing/signal_utils.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+def firwin_lowpass(highcut, fs, ntaps=DEFAULT_NUM_TAPS, window=DEFAULT_FILTER_WINDOW):
+    taps = firwin(ntaps, highcut,
+                  fs=fs, window=window, scale=False)
+    return taps
+
+
+def firwin_lowpass_filter(data, highcut, fs, ntaps=DEFAULT_NUM_TAPS, window=DEFAULT_FILTER_WINDOW):
+    taps = firwin_lowpass(highcut, fs, ntaps, window)
+    y = lfilter(taps, 1, data)
```
Comment (godsic) [inline _signal_processing/signal_utils.py:38]: Not sure, but I think filter's state needs to be propagated via `zi` argument.

File: _signal_processing/krakenSDR_signal_processor.py [hunk 3]
```diff
@@ -445,59 +441,61 @@ def run(self):
                                     freq_list.append(write_freq)
                                     doa_result_log_list.append(doa_result_log)
                                 else:
```
Comment (godsic) [inline _signal_processing/krakenSDR_signal_processor.py:444]: The data is saved only when DoA is not invoked (disabled, below squelch threshold or have no actives VFOs). Is it intentional?

Comment (redradist) [reply _signal_processing/krakenSDR_signal_processor.py:444]: Yes, it is collection of all DOA/IQ sample during long transaction

**Comment by krakenrf** [comment]
@redradist I can have a deeper look later, but any chance to help deconflict this with the latest pushes?

**Comment by redradist** [comment]
> @redradist I can have a deeper look later, but any chance to help deconflict this with the latest pushes?

@krakenrf I'll resolve conflicts today or tomorrow

File: _signal_processing/signal_utils.py (hunk 1 above)
Comment (redradist) [reply _signal_processing/signal_utils.py:70]: Actually, I played with it and have not noticed any differences

File: _signal_processing/signal_utils.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+import math
+
+import numpy as np
+from scipy.signal import firwin, lfilter, butter
+
+DEFAULT_NUM_OF_FREQ = 1024
+DEFAULT_FILTER_ORDER = 5
+DEFAULT_FILTER_WINDOW = 'hamming'
```
[the remark this answers is not in the document]
Comment (redradist) [reply _signal_processing/signal_utils.py:8]: I'll add an option

**Comment by redradist** [comment]
> @redradist I can have a deeper look later, but any chance to help deconflict this with the latest pushes?

@krakenrf I've update PR

**Comment by krakenrf** [comment]
Thanks will take a look soon

**Comment by krakenrf** [comment]
This is great, but I'll merge this into a new branch for now, as I think it still needs some work before we can incorporate it into the main.

1) The FM demod pulses quite a bit, I suppose that's an effect of the framing of IQ data into chunks. We've gotta figure out how to fix that
2) Need to add tooltips explaining the action of each new option
3) Need some way to remotely manage the wav files created. e.g. if a user runs on a headless Pi 4 the SD card could fill up very fast, rendering the system unable to boot after a short while. At the same time some way to listen to the wav files through a web interface would be good.
4) Need some way to handle long continuous signals (maybe just splitting into a wav file every 1 minute)
5) I ended up with a bug at one point, not sure what happened but the update rate became very slow until a restart. I think it was because it was on a long continuous signal for a few hours. But after switching away it never recovered.

**Comment by redradist** [comment]
> This is great, but I'll merge this into a new branch for now, as I think it still needs some work before we can incorporate it into the main.
> 
>     1. The FM demod pulses quite a bit, I suppose that's an effect of the framing of IQ data into chunks. We've gotta figure out how to fix that
> 
>     2. Need to add tooltips explaining the action of each new option
> 
>     3. Need some way to remotely manage the wav files created. e.g. if a user runs on a headless Pi 4 the SD card could fill up very fast, rendering the system unable to boot after a short while. At the same time some way to listen to the wav files through a web interface would be good.
> 
>     4. Need some way to handle long continuous signals (maybe just splitting into a wav file every 1 minute)
> 
>     5. I ended up with a bug at one point, not sure what happened but the update rate became very slow until a restart. I think it was because it was on a long continuous signal for a few hours. But after switching away it never recovered.

I thought to fix it in main branch those open questions, for some of them I already have solution like 1 minute threshold, which  I have not yet added in PR ...
Any way main branch is for development and leaving these changes in separate branch could lead to hard rebase in futureю
I prefer to have all this changes in main branch, anyway they are optional for user

**Comment by redradist** [comment]
> This is great, but I'll merge this into a new branch for now, as I think it still needs some work before we can incorporate it into the main.
> 
>     1. The FM demod pulses quite a bit, I suppose that's an effect of the framing of IQ data into chunks. We've gotta figure out how to fix that
> 
>     2. Need to add tooltips explaining the action of each new option
> 
>     3. Need some way to remotely manage the wav files created. e.g. if a user runs on a headless Pi 4 the SD card could fill up very fast, rendering the system unable to boot after a short while. At the same time some way to listen to the wav files through a web interface would be good.
> 
>     4. Need some way to handle long continuous signals (maybe just splitting into a wav file every 1 minute)
> 
>     5. I ended up with a bug at one point, not sure what happened but the update rate became very slow until a restart. I think it was because it was on a long continuous signal for a few hours. But after switching away it never recovered.

I thought to fix it in main branch those open questions, for some of them I already have solution like 1 minute threshold, which  I have not yet added in PR ...
Any way main branch is for development and leaving these changes in separate branch could lead to hard rebase in future.
I prefer to have all this changes in main branch, anyway they are optional for user

**Comment by krakenrf** [comment]
Hmm the main problem is that some users tend to click around and change settings randomly to see what they do. A user might activate the FM demod option, and not realize that it's quickly filling up the SD Card since it the GUI doesn't show anything changing. Then next time the unit will fail to boot because the SD card is too full. Thing like that result in a significant amount of support work for us. So I can't release to main a feature that does that, just yet at least.

In this case I might simply change the fm_demod to a development branch. All developers should PR to the dev branch first, and we only move things to main when they are 100% stable and user fool proof.

**Comment by krakenrf** [comment]
Another way we could do it is to just implement an "Enable Beta Features" button in the System Configuration box. So that the new features that could break user experience are hidden away from the GUI while in beta. Then everything can stay on the main branch.

**Comment by godsic** [comment]
I can suggest to maybe check the disk usage of the path IQ samples or WAV files are saved to, e. g., with https://www.geeksforgeeks.org/python-shutil-disk_usage-method/amp/. If, say, we are approaching the end of free space with some reasonable margin, then demodulation/IQ saving is aborted and warning message is displayed in the GUI suggesting to do some disks space cleaning. One might also introduce some garbage collection that removes files older then X days on every start.

-Mykola
________________________________

**Pull request merged**